### ISSUE TITLE
L025 removes alias which is used in subquery

### ISSUE THREAD

**Issue by sti0** [sqlfluff/sqlfluff#1462]
_If this is a parsing or linting issue, please include a minimal SQL example which reproduces the issue, along with the `sqlfluff parse` output, `sqlfluff lint` output and `sqlfluff fix` output when relevant._

## Expected Behaviour
With fixing ~L031~ L025 the alias should replaced correctly also in subqueries.

## Observed Behaviour
A alias from the parent query is used in a subquery. After fixing the file, the alias of the parent was removed while the alias within the subquery remains.

## Steps to Reproduce
Fix this script:
```
WITH base AS (
    SELECT
        nvl(col1, col2) AS col1,
        col2,
        col3,
        col4,
        col5
    FROM
        SRC_PRD.MDM_EXTRACT
),
without_dup AS (
    SELECT
        DISTINCT col1,
        col2,
        col3,
        col4
    FROM
        base
)
SELECT
    col1,
    col2,
    col3,
    col4,
    (
        SELECT
            count(*)
        FROM
            base b
        WHERE
            a.col2 = b.col2
            AND b.col5 = 'X'
    ) AS X_CNT,
    (
        SELECT
            count(*)
        FROM
            base b
        WHERE
            a.col2 = b.col2
            AND b.col5 = 'Y'
    ) AS Y_CNT,
    (
        SELECT
            count(*)
        FROM
            base b
        WHERE
            a.col2 = b.col2
            AND b.col5 = 'Z'
    ) AS Z_CNT,
FROM
    without_dup a;
```

Alias `a` from `without_dup` is removed but lives in the where clause of the subselects.

## Dialect
exasol

## Version
Include the output of `sqlfluff --version` along with your Python version
sqlfluff, version 0.6.6
Python 3.8.12
## Configuration
default

**Comment by tunetheweb** [comment]
I believe this is actually rule L025 not L031 which is causing the issue.

Minimal reproducible test case:

```sql
SELECT
    col1,
    (
        SELECT
            count(*)
        FROM
            base b
        WHERE
            a.col2 = b.col2
    )
FROM
    without_dup a;
```

And then run:
```
sqlfluff fix -f test.sql --rules L025 --fixed-suffix FIXED
diff test.sql testFIXED.sql
```

**Comment by adam-tokarski** [comment]
> I believe this is actually rule L025 not L031 which is causing the issue.

I'm not sure - yet - how it all works, but isn't that because of both these rules?
![image](https://user-images.githubusercontent.com/10441317/136008463-47930169-52e9-477e-9e36-5eab1748316d.png)

**Comment by barrywhart** [comment]
Quick note: There is a class `SelectCrawler` in `sqlfluff.core.rules.analysis.select_crawler` used by a couple of existing rules. It recursively walks through the queries and subqueries in a file. It could be useful for fixing this bug.

**Comment by mcannamela** [comment]
Here's a twist:
```
select 
    col_1, 
    col_2
from values(1, 2) as t(col_1, col_2)
```

```
root@eca08d9e0fd0:/dbt# sqlfluff fix test_fix.sql || echo exit $?
==== finding fixable violations ====
== [test_fix.sql] FAIL                                                                                                                                                          
L:   4 | P:  22 | L025 | Alias 't' is never used in SELECT statement.
==== fixing violations ====
1 fixable linting violations found
Are you sure you wish to attempt to fix these? [Y/n] ...
Attempting fixes...
Persisting Changes...
== [test_fix.sql] PASS
Done. Please check your files to confirm.
All Finished 📜 🎉!
```

```
select
    col_1,
    col_2
from values(1, 2)
```

dialect spark3
sqlfluff, version 0.9.0
Python 3.8.12

It correctly realizes that table alias `t` is unused, then the fix removes the column aliases as well, which are used. 

Is this actually a separate issue?

**Comment by barrywhart** [comment]
@mcannamela: That looks like a separate issue, because this one is talking about nested `SELECT` queries.

**Comment by mcannamela** [comment]
thanks, i'll raise it as such

**Comment by barrywhart** [comment]
Detailed design for a fix.
```
@dataclass
class AliasUsage:
    aliases: List[AliasInfo]
    tbl_refs: Set[str]

def _eval():
    if context.segment.is_type("select_statement"):
        select_info = get_select_statement_info(context.segment, context.dialect)
        if not select_info or not select_info.table_aliases:
            return None

        crawler = SelectCrawler(context.segment, context.dialect)
        query_alias_usage: Dict[Query, AliasUsage] = dict()
        _visit_query(crawler.query_tree, query_alias_usage, context.dialect)

        for alias in query_alias_usage[crawler.query_tree].aliases:
            if alias.aliased and alias.ref_str not in query_alias_usage[crawler.query_tree].tbl_refs:
                # Unused alias. Report and fix using existing L025 code.

def _visit_query(query: Query, query_alias_usage: Dict[Query, AliasUsage], dialect: Dialect):
    # Get aliases defined in query.
    for selectable in query.selectables:
        select_info = get_select_statement_info(selectable.selectable, context.dialect)
        query_alias_usage[query].aliases += select_info.table_aliases

    # Process references.
    for r in select_info.reference_buffer:
        for tr in r.extract_possible_references(
            level=r.ObjectReferenceLevel.TABLE
        ):
            # This function walks up the query's parent stack if necessary.
            # Reuses the existing "for r in references:" loop from L025.
            resolve_and_mark_reference(query, query_alias_usage, tr.part)

    # Visit children.
    for child in query.children:
        _visit_query(child, query_alias_usage, dialect)
```

**Comment by rileyrunnoe** [comment]
The bigquery dialect is also showing this error where the alias resolution becomes incorrect.

An aliased table in the overall selection but without a specific reference to that alias is removed, but the alias reference in the subquery still remains. This results in invalid bigquery sql.

Unfixed SQL:

```sql
SELECT col_1
, col_2
,col_3
from table_a AS a
WHERE NOT EXISTS (SELECT TRUE from table_b as b where a.col_4 = b.col_1)
```
"Fixed" SQL:

```sql
SELECT
    col_1
    , col_2
    , col_3
FROM table_a
WHERE NOT EXISTS (SELECT TRUE FROM table_b AS b WHERE a.col_4 = b.col_1)
```

Commented here as opposed to https://github.com/sqlfluff/sqlfluff/issues/2064 as it fits a little better.

**Comment by barrywhart** [comment]
@rileyrunnoe, @BI-MarcB, @adam-tokarski, @sti0: I have a PR for L025 which fixes this issue: #2273.
